pkg/api/types.go

```diff
@@ -1751,6 +1751,14 @@ func (w *WindowsProfile) GetWindowsSku() string {
 	return KubernetesDefaultWindowsSku
 }
 
+// GetSSHEnabled gets it ssh should be enabled for Windows nodes
+func (w *WindowsProfile) GetSSHEnabled() bool {
+	if w.SSHEnabled != nil {
+		return *w.SSHEnabled
+	}
+	return DefaultWindowsSSHEnabled
```

Comment:
This is probably not what we want. Are we assuming that this method is only invoked in execution flows after default enforcement flows have been traversed? (And if so, I still don’t think we want this method on this pointer object in this way.)

Reply:
This is following the same behavior as GetEnabledWindowsUpdate directly below.

This was done this way in order to not perform updates to the stored apimodel.json in upgrade/scale scenarios but still be able to enforce a default value.

What do you have in mind for this logic?

Reply:
So these methods are just for the E2E test runner?

Reply:
And to protection when operating on older templates.

Reply:
(I think - I initially wrote this ~2 months ago)

Reply:
Cool. So, I think the best way to do this to protect from future bugs (that method may be re-used in some critical cluster creation flow and then have unexpected results) is to do the default substitution in the test code itself, or maybe a a helper function in the e2e surface area that replaces nil values w/ defaults.

Reply:
Let me take a look at this again tomorrow and see if this can be cleaned up some.

Reply:
so I took a look and if there is an apimodel.json that does not have sshEnabled set is passed in to aks-engine generate the resulting apimodel.json in _output does persist the default value in code.

Reply:
The logic in defaults.go will set a value for w.SSHEnabled for generate operations but will not set a value for aks-engine upgrade or aks-engine scale operations - which looks like the behavior we want.
I see a unit test that verifies apimodel.json is not modified during those commands.
@jackfrancis i'm not following why you are against having a function first do a nil-check before returning a value - can you help me understand your concerns here?

Reply:
I think my main concern is that `GetSSHEnabled` is a accessor method that wraps the SSHEnabled bool pointer property. So I would expect a nil value of that to be falsy.

Reply:
This PR changed SSHEnabled property from a bool to a bool pointer so we could get an extra state - set to false, set to true, or unset by the user).
In the case where it is unset by the user wouldn't it make sense to set it to a default value instead of false?

Reply:
Right, perhaps I'm being too pedantic about that. It just seems that an accessor should return the truth, and if we want nil values to be semantically meaningful we should do that semantic enforcment so that it's sticky. I.e., so that by the time we use the accessors we can just have them be accessors, rather than being overloaded with that default value hijacking.

I think concretely what feels right to me is to set the default values during the defaults enforcement flow, and then confirm that the accessor is always being used only after that flow has been traversed and the data state representations have been rationalized (nil values converted to defaults, etc)